### PR TITLE
Fix PaliGemma Pad Token Masking During Training #35855

### LINKED ISSUE #35855: Paliegemma Pad Token not Masked
### System Info

- `transformers` version: 4.48.0
- Platform: Linux-6.1.92-3-x86_64-with-glibc2.31
- Python version: 3.10.15
- Huggingface_hub version: 0.26.2
- Safetensors version: 0.4.5
- Accelerate version: 1.1.1
- Accelerate config:    not found
- PyTorch version (GPU?): 2.5.1+cu124 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: parallel set-up
- Using GPU in script?: Yes
- GPU type: NVIDIA A100 80GB PCIe

### Who can help?
@amyeroberts @molbap
**Description:**

I encountered an unexpected behavior while using the Paligemma model. Specifically, the pad tokens' attention masks are being processed to zero (i.e., unmasked status) during the execution of the `self._update_causal_mask(attention_mask, token_type_ids, past_key_values, cache_position, input_ids, inputs_embeds, is_training)` function.

After investigating, I found that this behavior stems from a change in the code between the latest version (4.48.0) and the previous version (e.g., 4.43.0). The order in which the pad tokens are masked has been reversed, leading to inconsistent behavior, especially when using left padding.

### Code Comparison

**Latest Version (4.48.0):**
```python
if attention_mask is not None:
    causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
    mask_length = attention_mask.shape[-1]
    padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :].to(causal_mask.device)
    padding_mask = padding_mask == 0
    causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
        padding_mask, min_dtype
    )
    # we are training thus we need to create a full mask on the image + prefix but causal on suffix
    if is_training:
        causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
            token_type_ids[:, None, None, :].to(causal_mask.device) == 0, 0
        )
```

**Previous Version (4.43.0):**

```python
if token_type_ids is not None and labels is not None:
    # we are training thus we need to create a full mask on the image + prefix but causal on suffix
    target_length = cache_position[-1] + 1
    causal_mask = torch.full(
        (sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device
    )
    if sequence_length != 1:
        causal_mask = torch.triu(causal_mask, diagonal=1)
    causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
    causal_mask = causal_mask[None, None, :, :].expand(inputs_embeds.shape[0], 1, -1, -1)
    if attention_mask is not None:
        causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
        mask_length = attention_mask.shape[-1]
        padding_mask = causal_mask[:, :, :, :mask_length] + attention_mask[:, None, None, :].to(
            causal_mask.device
        )
        # unmask the prefill
        causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
            token_type_ids[:, None, None, :].to(causal_mask.device) == 0, 0
        )
        padding_mask = padding_mask == 0
        causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
            padding_mask, min_dtype
        )
```
**Issue Details**
In the latest version (4.48.0), the pad tokens are first masked using the padding_mask:

```python
padding_mask = padding_mask == 0
causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
    padding_mask, min_dtype
)
```
However, when the following block of code is executed during training:

```python
if is_training:
    causal_mask[:, :, :, :mask_length] = causal_mask[:, :, :, :mask_length].masked_fill(
        token_type_ids[:, None, None, :].to(causal_mask.device) == 0, 0
    )
```
the pad tokens (which have token_type_ids == 0) are unmasked again, effectively reverting the previous masking operation.

**Impact**
This behavior is particularly problematic when using left padding, as the pad tokens end up being unmasked in the latest version, whereas they were correctly masked in the previous version (4.43.0).

**Question**
Is this an intentional change, or is it a bug? If it’s a bug, could you please provide guidance on how to address it? I’d be happy to help with a fix if needed.

Thanks for your support!

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

```python
from PIL import Image
from transformers import AutoProcessor, PaliGemmaForConditionalGeneration

processor = AutoProcessor.from_pretrained("path")
model = PaliGemmaForConditionalGeneration.from_pretrained("path")

prompt = ["<image>caption en","<image>caption en wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww wwwwwwwwwwwwwwwwwww"]
labels = ["123123123123","456456456456"]
image_file = ["image path"] * 2
raw_image = [Image.open(i).convert("RGB") for i in image_file]

inputs = processor(text=prompt, images=raw_image, suffix=labels,
                    return_tensors="pt", padding="longest")
model(**inputs)
```

### Expected behavior

The pad tokens end up being unmasked.

### PR DESCRIPTION
**Pull request by sambhavnoobcoder**
## Problem Statement
In PaliGemma model's `_update_causal_mask` function, padding tokens were being incorrectly unmasked during training mode. This occurred because the order of operations first applied padding masks and then unmasked prefix tokens (including pad tokens) during training, leading to inconsistent behavior especially with left padding.

Fixes : #35855 

## Approach
After analyzing the masking logic, we identified that the issue stemmed from the sequence of operations in mask application. The solution was to reorder the masking operations to:
1. First unmask prefix tokens during training mode
2. Then apply padding masks

This ensures pad tokens remain masked regardless of their position or training state.

## Implementation
The fix involved reordering the masking operations in the `_update_causal_mask` function while maintaining the same mathematical operations. This approach ensures:
- No changes to the underlying logic
- Minimal code changes
- Preserved backward compatibility
- Consistent behavior across training and inference modes

### Test Coverage
The new test validates:
1. Pad tokens remain masked (value = dtype.min) during training
2. Non-pad tokens are properly masked/unmasked based on their position
3. Behavior is consistent across different batch sizes and sequence lengths

### Screenshots

Screenshot 2025-01-24 at 12 57 47 AM

**cc:** @amyeroberts @molbap @zucchini-nlp  kindly review this whenever you find the time .

### CODE REVIEW

File: tests/models/paligemma/test_attention_mask.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import unittest
```
Comment (zucchini-nlp) [inline tests/models/paligemma/test_attention_mask.py:1]: We need to add a test in `test_modeling_paligemma.py` file with tiny dummy model weights, and a single test covering attention with/without suffix should be enough

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for opening a PR! Can you move the test to the `test_modeling_paligemma` and adapt it for dummy weights. I don't think we need slow test with big model, since we're just checking mask on attn weights

File: tests/models/paligemma/test_attention_mask.py (hunk 1 above)
Comment (sambhavnoobcoder) [reply tests/models/paligemma/test_attention_mask.py]: okay . I have done this in [96e1b43](https://github.com/huggingface/transformers/pull/35859/commits/96e1b43f4a04c88c08b26be1c75b53d5f8968ff1) commit , and removed the separate testing file from the PR . kindly review it for any more changes as well . if everything is fine , I'll push the fixed style changes and then we are good to merge .

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for iterating on this!

LGTM, but for the test we better just add a new one under `PaliGemmaForConditionalGenerationModelTest` which already creates dummy model and inputs. Sorry if it wasn't clear the first time
Also, let's test the model forward in general with `output_attentions=True`, instead of relying on a single `update_causal_mask` call. And add test if token type ids are passed vs when not passed, to make sure attention is masked correctly in all cases

**Comment by sambhavnoobcoder** [comment]
Thanks for the feedback! I've moved the test into `PaliGemmaForConditionalGenerationModelTest` class and am now using the existing dummy model setup instead of creating a separate test class. Instead of directly testing `_update_causal_mask` , I've switched to testing the full model forward pass with `output_attentions=True` which better reflects real usage. The test now covers both scenarios - with and without token_type_ids - to ensure attention masking works correctly in all cases. It verifies that pad tokens have zero attention weights across all attention layers and that the masking behavior remains consistent regardless of token type presence. Let me know if you'd like me to make any additional adjustments!

P.s also made the style fixup to make this ready to merge .

**Review by zucchini-nlp — APPROVED** [review]
Perfect, thanks a lot! Btw, you can remove the `PaliGemmaAttentionMaskTest` as it is testing the same thing

Requesting core maintainer's review and we'll merge

**Comment by sambhavnoobcoder** [comment]
Thank you for the review @zucchini-nlp . I have removed `PaliGemmaAttentionMaskTest` in [a015a90](https://github.com/huggingface/transformers/pull/35859/commits/a015a90b12f5d3d524bbef20ba10cb2644edfc51) commit .


**cc:** @ArthurZucker for final review / merging.

**Comment by molbap** [comment]
Thanks a lot! gentle ping @ArthurZucker for this! 
I'm sorry for not looking earlier, I thought we already moved that method around within `_update_causal_mask`. It turns out we are calling this method once more in `prepare_inputs_for_generation` when it's a PaliGemma2 model with HybridCache, any possible interaction with the training?

**Comment by sambhavnoobcoder** [comment]
Thank you for the careful review @molbap  ! I've double-checked the interaction between `_update_causal_mask` and its usage in `prepare_inputs_for_generation` with HybridCache.

The current implementation maintains correct masking behavior in both contexts because:

1. The reordered masking operations (first prefix unmasking, then padding masks) ensure pad tokens stay masked regardless of where the method is called from
2. The `is_training` flag is properly propagated in both cases
3. Our test coverage in `test_attention_mask_with_token_types` verifies the masking behavior works correctly with and without token_type_ids

So the masking logic remains consistent whether called during regular training/inference or during generation with HybridCache, so no additional changes should be needed.

**Review by ArthurZucker — APPROVED** [review]
Cool! Thanks for fixing 🤗
